### DISCUSSION TITLE
RootCNC v3 - update and Test

### DISCUSSION THREAD

**Discussion by NEWTech-Creative** [grblHAL/ESP32#102] [General]
I would like to update and text the RootCNC v3 board and grblHAL. I have found some errors on existing board map and would like to update and text via the Web builder. How can I make changes and provide feedback with testing? Thank you

[root_cnc_v3_map.txt](https://github.com/grblHAL/core/files/14548409/root_cnc_v3_map.txt)

**Comment by terjeio** [comment]
I have now updated the Web Builder source with your map, but not the github repo.

**Reply by NEWTech-Creative** [reply]
Thank you for your help. I am wanting to test a prototype board called the root CNC Pro board. Here is the map file and board info file. Can you please add to the web builder to test? Thank you
[ESP32_ROOTCNC_PRO.json](https://github.com/grblHAL/ESP32/files/15038134/ESP32_ROOTCNC_PRO.json)
[root_cnc_pro_map.txt](https://github.com/grblHAL/ESP32/files/15038145/root_cnc_pro_map.txt)

is there a better way I can request this? 

Thank you

**Reply by terjeio** [reply]
[Added](https://github.com/grblHAL/ESP32/commit/c36e32f03c753ae39ab2c72bbb3034d19cf08935).

> is there a better way I can request this?

The correct way is to compile locally and test, then create a pull request with the changes.

**Reply by NEWTech-Creative** [reply]
Thank you! I appreciate your guidance. I attempted to build on the web builder and had an error 422. I will attempt to use visual studio to compile and test.

**Reply by terjeio** [reply]
> I attempted to build on the web builder and had an error 422. 

Fixed now.

**Comment by NEWTech-Creative** [comment]
I have have a further investigation into this pro board. the board has 3x tmc_5160's onboard with a cs pin. Does grblHAL have the capability of using a cs pin for these drivers? here is my amended map. let me know if I need to define an alternate way? Thank you
[root_cnc_pro_map.txt](https://github.com/grblHAL/ESP32/files/15046937/root_cnc_pro_map.txt)

**Reply by terjeio** [reply]
> the board has 3x tmc_5160's onboard with a cs pin.

What about motors M3 - M5 then, they need external drivers?

> Does grblHAL have the capability of using a cs pin for these drivers?

Yes. `MOTOR_CS_PIN` has to be [defined](https://github.com/grblHAL/ESP32/blob/1e5aef8a3426db1bf3370ed1912461ec203d807e/main/boards/xPro_v5_map.h#L105) for that.

**Reply by NEWTech-Creative** [reply]
I will test with my compiled version and send through when successful.

**Comment by NEWTech-Creative** [comment]
Thank you for your ongoing support. I am testing both the RootCNC v3 and Pro boards. I wanted to clarify how to setup auto squared via the webbuilder especially with the RootCNC v3. When I select 'auto-squared' on the y axis I get an error 'AUTO SQUARE: available 0, required 1.'. I am unsure what this means as I know the pins are mapped correctly. Can you please clarify what this error is and how I can amend it? 
My second question is how can I setup a rotary A axis? Do I just add as a 4 axis machine and it will default to the A axis? how does this affect selecting what axis is the Y2 axis and what is the rotary axis? 
Is there a more detailed page regarding axis setup? I have tried to search in wiki and didn't seem to find info.
Thank you!

**Reply by terjeio** [reply]
> When I select 'auto-squared' on the y axis I get an error 'AUTO SQUARE: available 0, required 1.'. I am unsure what this means as I know the pins are mapped correctly.

Auto squaring requires a second limit input for the axis, which is available for v2 and v3 but not the Pro?
To make auto-squaring available in the Web Builder the capability has to be declared in the [driver.json](https://github.com/grblHAL/ESP32/blob/master/driver.json) file, this is missing from the v2 and v3 entries. I'll add it in the next commit.

> My second question is how can I setup a rotary A axis? Do I just add as a 4 axis machine and it will default to the A axis?

Yes.

> how does this affect selecting what axis is the Y2 axis and what is the rotary axis?

_Motor_ selection is done automatically, The assigned Y2 motor is linked to the Y axis except during homing when the Y motors will become unlinked, when needed, to complete the auto squaring sequence. 
The automatic selection assigns M3+ motors to additional axes and ganged/autosquared motors from the remaining unassigned motors in the map file.

> Is there a more detailed page regarding axis setup?

No, I have not written any.

**Reply by NEWTech-Creative** [reply]
Thank you for your clarity here! That make sense :)  I've been working with the v3 for the movement and is working great. I'll check the auto squaring function before marking off as tested. 
Is there a direct contact email or discord that I can reach you on terjeio? my email is newtechcreative@gmail.com
